### PR TITLE
[FLINK-5963] [table] Remove prepare mapper of DataSetAggregate.

### PR DESCRIPTION
**Pull request by fhueske**
We remove the preparation mapper to 
- save one operator
- be able to apply `AggregateFunction.accumulate()` in a `GroupCombineFunction` or `MapPartitionFunction` for pre-aggregation or in a `GroupReduceFunction` for final aggregation.

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetAggregate.scala [hunk 1]
```diff
@@ -87,47 +89,67 @@ class DataSetAggregate(
...
-        .name(aggOpName)
+      if (preAgg.isDefined) {
+        inputDS
+          // pre-aggregation
+          .groupBy(grouping: _*)
+          .combineGroup(preAgg.get)
+          .returns(preAggType.get)
+          .name(aggOpName)
+          // final aggregation
+          .groupBy(grouping.indices: _*)
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetAggregate.scala]: I played around this PR with different test modes, everything works very well. Just curious why you use grouping.indices as the grouping key here?

Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetAggregate.scala]: The pre-aggregation function modifies the schema of rows. It puts all grouping keys first, followed by all accumulators. Therefore, the following final aggregation needs to group on the first `n` fields.
Before, the prepare mapper changed the layout.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetPreAggFunction.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  * @param aggregates The aggregate functions.
+  * @param aggInFields The positions of the aggregation input fields.
+  * @param groupingKeys The positions of the grouping keys in the input.
+  */
+class DataSetPreAggFunction(
+    private val aggregates: Array[AggregateFunction[_ <: Any]],
+    private val aggInFields: Array[Int],
+    private val groupingKeys: Array[Int])
+  extends AbstractRichFunction
+    with GroupCombineFunction[Row, Row]
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetPreAggFunction.scala]: Can we align with and extend? 
The suggestion just ref to: [Scalar Class/Object/Trait constructors declarations ](http://docs.scala-lang.org/style/declarations)

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetAggFunction.scala [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  private var intermediateGKeys: Option[Array[Int]] = None
+  private val aggsWithIdx: Array[(AggregateFunction[_], Int)] = aggregates.zipWithIndex
+  private var accumulators: Array[Accumulator] = _
+
+  override def open(config: Configuration) {
+    Preconditions.checkNotNull(aggregates)
+    Preconditions.checkNotNull(aggInFields)
+    Preconditions.checkNotNull(aggOutMapping)
+    Preconditions.checkNotNull(gkeyOutMapping)
+    accumulators = new Array(aggregates.length)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetAggFunction.scala]: good point! thanks

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetAggregate.scala [hunk 4]
```diff
@@ -87,47 +89,67 @@ class DataSetAggregate(
...
-
-    val groupReduceFunction = AggregateUtil.createAggregateGroupReduceFunction(
-      namedAggregates,
-      inputType,
-      rowRelDataType,
-      grouping,
-      inGroupingSet)
+    val (preAgg: Option[DataSetPreAggFunction],
+          preAggType: Option[TypeInformation[Row]],
+          finalAgg: GroupReduceFunction[Row, Row]) =
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetAggregate.scala]: OK

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetPreAggFunction.scala (hunk 2 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetPreAggFunction.scala]: OK

**Comment by fhueske** [comment]
Hi @sunjincheng121, thanks for your review. I addressed you comments. 
Cheers, Fabian

**Comment by fhueske** [comment]
Thanks for the reviews @shaoxuan-wang and @sunjincheng121.
Will merge this PR

**Comment by twalthr** [comment]
Thanks @fhueske. I will rebase this and replace foreach with while loops, and merge it.

**Pull request closed without merging**